File: urbanairship-core/src/main/java/com/urbanairship/util/CachedValue.java
/* Copyright Airship and Contributors */

package com.urbanairship.util;


import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;
import androidx.core.util.Predicate;

/**
 * Caches a value in memory with an expiration.
 *
 * @param <T>
 *
 * @hide
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
public class CachedValue<T> {

    private final Object lock = new Object();
    private final Clock clock;
    private long expiration;
    private T value;

    public CachedValue() {
        this(Clock.DEFAULT_CLOCK);
    }

    public CachedValue(@NonNull Clock clock) {
        this.clock = clock;
    }
    public void set(@Nullable T value, long expiryDateMs) {
        synchronized (lock) {
            this.value = value;
            this.expiration = expiryDateMs;
        }
    }

    public void expire() {
        synchronized (lock) {
            this.value = null;
            this.expiration = 0;
        }
    }

    public void expireIf(Predicate<T> predicate) {
        synchronized (lock) {
            if (value != null && predicate.test(value)) {
                this.value = null;
                this.expiration = 0;
            }
        }
    }

    public long remainingCacheTimeMillis() {
        long remaining = expiration - clock.currentTimeMillis();
        if (remaining >= 0) {
            return remaining;
        }
        return 0;
    }

    @Nullable
    public T get() {
        synchronized (lock) {
            if (clock.currentTimeMillis() >= expiration) {
                return null;
            }

            return value;
        }
    }

}

File: urbanairship-core/src/main/java/com/urbanairship/util/CachedValue.kt
/* Copyright Airship and Contributors */
package com.urbanairship.util

import androidx.core.util.Predicate
import java.time.Duration as JavaDuration
import java.time.Instant
import kotlin.time.Duration
import kotlin.time.toKotlinDuration

/**
 * Caches a value in memory with an expiration.
 *
 * @param <T>
 *
 * @hide
</T> */
internal class CachedValue<T> (
    private val clock: Clock = Clock.DEFAULT_CLOCK
) {

    private val lock = Any()
    private var expiration: Instant = Instant.EPOCH
    private var value: T? = null

    fun set(value: T?, expiresAt: Instant) {
        synchronized(lock) {
            this.value = value
            this.expiration = expiresAt
        }
    }

    fun expire() {
        synchronized(lock) {
            this.value = null
            this.expiration = Instant.EPOCH
        }
    }

    fun expireIf(predicate: Predicate<T>) {
        synchronized(lock) {
            if (value == null) { return@synchronized }
            if (!predicate.test(value)) { return@synchronized }

            this.value = null
            this.expiration = Instant.EPOCH
        }
    }

    fun remainingCacheTime(): Duration {
        val remaining = JavaDuration.between(clock.now(), expiration)
        return if (remaining.isNegative) Duration.ZERO else remaining.toKotlinDuration()
    }

    fun get(): T? {
        synchronized(lock) {
            if (remainingCacheTime() > Duration.ZERO) {
                return value
            }

            value = null
            return value
        }
    }
}
